adapter/src/main/java/com/twilio/kudu/dataloader/DataLoader.java

```diff
@@ -83,16 +85,17 @@ public DataLoader(final String url, final Scenario scenario, Integer threadPoolS
     timestampGenerator.getColumnValue();
     this.scenarioStartTimestamp = timestampGenerator.minValue;
     this.scenarioEndTimestamp = timestampGenerator.maxValue;
-    // pick the first cube to determine the FLOOR mod value
-    // TODO figure out how to handle a mix of different cube time rollup
-    // granularities
-    this.cubeGranularityFloorMod = calciteKuduTable.getCubeTables().isEmpty() ? DateTimeUtils.MILLIS_PER_SECOND
-        : ((CalciteModifiableKuduTable) calciteKuduTable.getCubeTables().get(0)).getCubeMaintainer().getFloorMod();
+    // Use the largest cube granularity to determine the total number of tasks that
+    // can be
+    // parallelized. Eg. if the scenario time range is 14 days and we have 1 hourly
+    // and 1 daily
+    // cube, we can run 14 tasks in parallel (one for each day).
+    this.cubeGranularityFloorMod = getMaxCubeGranularity();
     int numTasks = (int) Math.ceil((float) (scenarioEndTimestamp - scenarioStartTimestamp) / cubeGranularityFloorMod);
 
     // set thread pool to MIN(2 * number of cpus, numTasks) by default
-    this.threadPoolSize = threadPoolSize != null ? threadPoolSize
-        : Math.min(Runtime.getRuntime().availableProcessors() * 2, numTasks);
+    this.threadPoolSize = Math
+        .min(threadPoolSize != null ? threadPoolSize : Runtime.getRuntime().availableProcessors() * 2, numTasks);
```

Comment:
nit: Just my fingers burning over this availableProcessors logic says we can go beyond it as hyperthreading in a JVM is weird and can have more threads running than the number of processors.
Feel free to experiment and remove to just use "numTasks" as JVM goes well beyond it just fine.

Reply:
I modified the default thread pool size to be `Runtime.getRuntime().availableProcessors() * 10` since numTasks could potentially be quite large.